project_recalculate/models/project_task.py

```diff
@@ -0,0 +1,168 @@
+# -*- coding: utf-8 -*-
+##############################################################################
+#
+#    OpenERP, Open Source Management Solution
+#    This module copyright :
+#        (c) 2014 Antiun Ingenieria, SL (Madrid, Spain, http://www.antiun.com)
+#                 Endika Iglesias <endikaig@antiun.com>
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp import models, fields, api, _
+from openerp.exceptions import Warning
+from datetime import timedelta
+import pytz
+
+
+class ProjectTask(models.Model):
+    _inherit = 'project.task'
+
+    from_days = fields.Integer(
+        string='From days',
+        help='Anticipation days from date begin or date end', default=0)
+    estimated_days = fields.Integer(
+        string='Estimated days', help='Estimated days to end', default=1)
+
+    @api.one
+    @api.constrains('estimated_days')
+    def _estimated_days_check(self):
+        if not self.estimated_days > 0:
+            raise Warning(_('Estimated days must be greater than 0.'))
+
+    def _count_days_without_weekend(self, date_start, date_end):
+        days = (date_end - date_start).days
+        return sum(1 for x in xrange(days)
+                   if (date_start + timedelta(x)).weekday() < 5)
+
+    def _count_days_weekend(self, date_start, date_end):
+        days = (date_end - date_start).days
+        return sum(1 for x in xrange(days)
+                   if (date_start + timedelta(x)).weekday() >= 5)
+
+    def _correct_days_to_workable(self, date, increment=True):
+        while date.weekday() >= 5:
+            if increment:
+                date += timedelta(days=1)
+            else:
+                date -= timedelta(days=1)
+        return date
+
+    def _calculate_date_without_weekend(self, date_start, days,
+                                        increment=True):
+        total_days = 0
```

Comment:
This code is tremendously ugly and have a lot of lines for a simple function like that. It has to be totally refactored with the remarks I have made in the other method.